### PR TITLE
PUG's builtin composition as default + static rendering strategy

### PR DESCRIPTION
**Pull request by pioneerskies**
## Revieweing w/o being cluttered by deps update

@AleOnRails @danielegiorgis as you can see this PR is "cluttered" by two things:

* vendor dependencies updates
* documentation updates

**[Here](https://github.com/welaika/wordless/pull/257/files/a4268e7391934a85bf478bdba4d39df1ea7659ed..313186cccfc190c4a764b79a4afab9d055bc610e?file-filters%5B%5D=.json&file-filters%5B%5D=.php&file-filters%5B%5D=.pug) you can review the declutterd new implementations to the code**

This way you should be able to review the work having a clearer vision about what happened in the code :)

## Updated documentation

[Here](https://github.com/welaika/wordless/pull/257/files/a4268e7391934a85bf478bdba4d39df1ea7659ed..313186cccfc190c4a764b79a4afab9d055bc610e?file-filters%5B%5D=.rst)  are **only documentation** updates (but they are a vivid work in progress)

Here we have the new documentation preview: https://wordless.readthedocs.io/en/static_rendering/

This is a long running work. Thus I propose to disjoin the completion of all the documentation and its grammar review from the DoD of this PR. I'll continue documenting directly on master if we're able to approve and merge this one quickly.

## Features from other sub-PRs

* https://github.com/welaika/wordless/pull/259
* https://github.com/welaika/wordless/pull/260

These PRs was already reviewed, thus I do not expect to be blocked into this one for the merged code from those. Unless problems found will be about integration between the PRs.

### Components

I've recognized some of our internal recurring practices and decided to ship them by default in Wordless. One is component based templating. in the new __ecosystem__ "components" are written as PUG's `mixin`s.

We also ship a new class called `Wordless\Component` to build `Component` objects with a straight, declarative and validated data structure. Giving a `Wordless\Component` object as parameters to a pug `component` (or `mixin`: they are synonims in Wordless now) in order to be rendered has the utility to ensures a couple of things:

* move away a lot of logic from views
* objects are validated when created: the dev has the simplestester® interface to deal with
* validation are built upon Symphony's validation framework. No more wheels reinventing.
* all is declarative: dev is required to create a new object, inside the object every attribute is declared, inside the component view you can __get__ only declared properties...all under the sun's light. I like it, actually

In order to dry a bit this description, I invite you to read the _in progress_ documentation @ https://wordless.readthedocs.io/en/static_rendering/_pages/usage/rendering_pug.html#components in order to get the whole scenario.

#### Validation errors in production

By default no exception will be raised in production/staging, while in development a validation error will rise an ad hox exception. An hook is introduced in order to require custom action when in development. [here](https://github.com/welaika/wordless/pull/257/files/a4268e7391934a85bf478bdba4d39df1ea7659ed..313186cccfc190c4a764b79a4afab9d055bc610e?file-filters%5B%5D=.json&file-filters%5B%5D=.php&file-filters%5B%5D=.pug#diff-c28c2e2721937f4231cf6aae49dc3911d5cf74be9432efacafc49b9cdd007ee8R58) is the code.

### Full PUG templates

Wordless has always used an hybrid approach between custom php partial rendering and PUG. From now on, even being fully backwards compatible with older themes, we'll completely rely on PUG. Having the only `render_template()` put into `index.php`, we'll use `extend`/`include`/`mixin` functionality directly from PUG.

The benefit it terms of performance is - I mean - HUGE. 1/3 rendering time in the worst scenario and tons less memory usage.

I've almost completed the documentation about the new common practice in the _in progress_ documentation @ https://wordless.readthedocs.io/en/static_rendering/_pages/usage/rendering_pug.html where you can see through examples the new approach. (be aware: some code snippets point to wrong lines after some recent merges. I'm sorry for that and I'll fix it after this PR will be closed).

Previously used render helpers are always there, but they are now mainly intended to handle the need of writing plain PHP templates. Which is never our internal case. Documentation has been moved into an appendix @ https://wordless.readthedocs.io/en/static_rendering/_pages/appendix/plainphp.html (but do not waste time reading it now)

### Static render

We introduce a new render helper called `render_static()`. When used to render a PUG file (template or partial) it will generate a static HTML file on the first call and will serve it directly. The HTML will be cached for the next call and until manually deleted using an ad hoc backend function put into a new custom WP menu.

We can transform in static HTML and entire site, an entire template of just the smallest (and maybe slowest :P) piece of pug we have into a theme.

Documentation has not yet been written.

The static cache could be bypassed by setting a `BYPASS_STATIC` constant to `true` in `wp-config.php`.

### Updated vanilla theme

The vanilla theme is updated to reflect and exemplify new features.

### CODE REVIEW

File: wordless/helpers/render_helper.php [hunk 1]
```diff
@@ -72,13 +86,40 @@ function render_template($name, $locals = array()) {
...
+                        $bypass_static = var_export(BYPASS_STATIC, true); // constant could be a boolean so we uniform to a string representation
+                    } else {
+                        $bypass_static = 'false'; // default value
+                    }
+
                     if ( in_array( $env, array('staging', 'production') ) ) {
-                        \Pug\Optimizer::call(
-                            'displayFile', [$template_path, $locals], WordlessPugOptions::get_options()
-                        );
+                        if (true === $static && false == $bypass_static) {
```
Comment (pioneerskies) [inline wordless/helpers/render_helper.php:99]: `false` here i a boolean and it will never work like that. Ouch.

Comment (pioneerskies) [reply wordless/helpers/render_helper.php:99]: Reminder: do more tests in `staging`/`production` environment

Comment (pioneerskies) [reply wordless/helpers/render_helper.php]: > false is a boolean and it will never work like that. Ouch.

fixed

**Comment by pioneerskies** [comment]
Cleaned up and reordered commit history

File: rtfm/source/_pages/build/prod.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    yarn clean:dist && yarn build:prod
+
+Production build will essentially:
+
+* enable Webpack's `production mode`_
+* do not produce source maps for CSS
+* do minimize assets
+
+.. note::
+    By default the production build **will** produce source-maps for JS; this is done to
```
Comment (danielegiorgis) [inline rtfm/source/_pages/build/prod.rst:17]: with the commit https://github.com/welaika/wordless/pull/261/commits/6b95f16c66fc6541d5a955a20bcedf7abee63dc0
we need to change this part

File: rtfm/source/_pages/usage/css_js.rst [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+They will be compiled and resulting compilation files will be moved in the corresponding
+``assets/xxx`` folder.
+
+Compilation, naming and other logic is fully handled by webpack.
+
+Images will be optimized by `ImageminPlugin`_. The default setup already translates
+``url`` s inside css/scss files in order to point to images in the
+right folder.
+
+.. _ImageminPlugin: https://www.npmjs.com/package/imagemin-webpack-plugin
```
Comment (danielegiorgis) [inline rtfm/source/_pages/usage/css_js.rst:39]: as before the new link is https://github.com/webpack-contrib/image-minimizer-webpack-plugin

File: rtfm/source/_pages/usage/css_js.rst [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+They must be placed inside ``src/javascript/`` and
+``src/stylesheets/`` and ``src/images/``.
+
+They will be compiled and resulting compilation files will be moved in the corresponding
+``assets/xxx`` folder.
+
+Compilation, naming and other logic is fully handled by webpack.
+
+Images will be optimized by `ImageminPlugin`_. The default setup already translates
```
Comment (danielegiorgis) [inline rtfm/source/_pages/usage/css_js.rst:35]: change with the the new imagemin

File: rtfm/source/_pages/usage/initializers.rst [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+one with a specific target:
+::
+
+  config/initializers
+  ├──── backend.php
+  ├──── custom_gutenberg_acf_blocks.php
+  ├──── custom_post_types.php
+  ├──── default_hooks.php
+  ├──── hooks.php
+  ├──── login_template.php
```
Comment (danielegiorgis) [inline rtfm/source/_pages/usage/initializers.rst:23]: Perhaps useless comment. the description below is missing

File: wordless/helpers/component_helper.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    private $validator = null;
+
+    function __construct(array $locals) {
+        $this->validator = Validation::createValidatorBuilder()->addMethodMapping('loadValidatorMetadata')->getValidator();
+        $this->locals = $locals;
+        try {
+            $this->setProperties();
+            $this->validate();
+        } catch (ComponentValidationException $e) {
+            if ( 'production' === ENVIRONMENT ) {
```
Comment (danielegiorgis) [inline wordless/helpers/component_helper.php:56]: In the PR's description you say "Validation errors in production
By default no exception will be raised in production/staging"

maybe i'm wrong, but in this way staging acts as if it's development

File: rtfm/source/_pages/usage/initializers.rst (hunk 5 above)
Comment (pioneerskies) [reply rtfm/source/_pages/usage/initializers.rst:23]: Going to push an update

File: rtfm/source/_pages/usage/css_js.rst (hunk 4 above)
Comment (pioneerskies) [reply rtfm/source/_pages/usage/css_js.rst:35]: Going to push updates

File: wordless/helpers/component_helper.php (hunk 6 above)
Comment (pioneerskies) [reply wordless/helpers/component_helper.php:56]: Yep, you're right: we have `staging` as possible value elsewhere in the code and in the docs. Sweet catch!

Comment (pioneerskies) [reply wordless/helpers/component_helper.php:56]: BTW the problem is in teh description: I actually **want** to drop exception when on staging. I think it is so important. So the code is ok and my description inaccurate.

I'm going to resolve this one! Thanks again.

File: rtfm/source/_pages/build/prod.rst (hunk 2 above)
Comment (pioneerskies) [reply rtfm/source/_pages/build/prod.rst:17]: Updating the doc to

```rst
.. note::
    By default the production build won't produce source-maps for JS.

    You can easily change this behaviour updating `` const needSourceMap = (env.DEBUG == 'true');``
    to ``  const needSourceMap = true;`` in ``webpack.env.js``
```

**Pull request merged**